File: source/Java/GlobalVariables.java
/*	Class:		GlobalVariables
 *	Purpose:	To hold the global variables
 *	Version:	Thinknowlogy 2018r1 (ShangDi 上帝)
 *************************************************************************/
/*	Copyright (C) 2009-2018, Menno Mafait. Your suggestions, modifications,
 *	corrections and bug reports are welcome at http://mafait.org/contact/
 *************************************************************************/
/*	This program is free software: you can redistribute it and/or modify
 *	it under the terms of the GNU General Public License as published by
 *	the Free Software Foundation, either version 2 of the License, or
 *	(at your option) any later version.
 *
 *	This program is distributed in the hope that it will be useful,
 *	but WITHOUT ANY WARRANTY; without even the implied warranty of
 *	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *	GNU General Public License for more details.
 *
 *	You should have received a copy of the GNU General Public License along
 *	with this program; if not, write to the Free Software Foundation, Inc.,
 *	51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *************************************************************************/

class GlobalVariables
	{
	protected static boolean hasDisplayedIntegrityWarning;
	protected static boolean hasDisplayedMessage;
	protected static boolean hasDisplayedWarning;
	protected static boolean hasFoundAnswerToQuestion;
	protected static boolean hasFoundQuery;
	protected static boolean isAssignmentChanged;
	protected static boolean isFirstAnswerToQuestion;
	protected static boolean isQuestionAlreadyAnswered;

	protected static byte result;

	protected static short currentAssignmentLevel;
	protected static short currentLanguageNr;
	protected static short currentUserNr;
	protected static short currentWriteLevel;

	protected static int currentSentenceNr;
	protected static int currentSentenceItemNr;

	protected static int nActiveQueryItems;
	protected static int nInactiveQueryItems;
	protected static int nArchivedQueryItems;
	protected static int nReplacedQueryItems;
	protected static int nDeletedQueryItems;
	protected static int nDeletedItems;

	protected static int nUserGeneralizationWords;
	protected static int nUserSpecificationWords;
	protected static int nUserRelationWords;

	protected static int removeSentenceNr;
	protected static int removeStartItemNr;

	protected static long cleaningTime;
	protected static long openFileTime;
	protected static long selectionExecutionTime;
	protected static long writingTime;

	protected static SelectionList adminConditionList;
	protected static SelectionList adminActionList;
	protected static SelectionList adminAlternativeList;

	protected static WordItem currentLanguageWordItem;
	protected static WordItem firstAssignmentWordItem;
	protected static WordItem firstCollectionWordItem;
	protected static WordItem firstContextWordItem;
	protected static WordItem firstPossessiveNounWordItem;
	protected static WordItem firstPredefinedWordItem;
	protected static WordItem firstSpecificationWordItem;
	protected static WordItem firstTouchedWordItem;
	protected static WordItem firstUserProperNounWordItem;
	protected static WordItem firstWordItem;
	protected static WordItem lastCollectionWordItem;
	protected static WordItem lastContextWordItem;
	protected static WordItem lastPredefinedWordItem;
	protected static WordItem predefinedNounLanguageWordItem;
	protected static WordItem predefinedNounUserWordItem;

	protected static StringBuffer currentPathStringBuffer;
	protected static StringBuffer interfaceLanguageStringBuffer;
	protected static StringBuffer learnedFromUserStringBuffer;
	protected static StringBuffer queryStringBuffer;
	protected static StringBuffer readUserSentenceStringBuffer;
	protected static StringBuffer writtenSentenceStringBuffer;
	protected static StringBuffer writtenUserSentenceStringBuffer;


	// Protected methods

	protected static void initialize()
		{
		hasDisplayedIntegrityWarning = false;
		hasDisplayedMessage = false;
		hasDisplayedWarning = false;
		hasFoundAnswerToQuestion = false;
		hasFoundQuery = false;
		isAssignmentChanged = false;
		isFirstAnswerToQuestion = false;
		isQuestionAlreadyAnswered = false;

		result = Constants.RESULT_OK;

		currentAssignmentLevel = Constants.NO_ASSIGNMENT_LEVEL;
		currentLanguageNr = Constants.NO_LANGUAGE_NR;
		currentUserNr = Constants.NO_USER_NR;
		currentWriteLevel = Constants.NO_WRITE_LEVEL;

		currentSentenceNr = 1;	// First sentence
		currentSentenceItemNr = Constants.NO_ITEM_NR;

		nActiveQueryItems = 0;
		nInactiveQueryItems = 0;
		nArchivedQueryItems = 0;
		nReplacedQueryItems = 0;
		nDeletedQueryItems = 0;
		nDeletedItems = 0;

		nUserGeneralizationWords = 0;
		nUserSpecificationWords = 0;
		nUserRelationWords = 0;

		removeSentenceNr = Constants.NO_SENTENCE_NR;
		removeStartItemNr = Constants.NO_ITEM_NR;

		cleaningTime = 0;
		openFileTime = 0;
		selectionExecutionTime = 0;
		writingTime = 0;

		adminConditionList = null;
		adminActionList = null;
		adminAlternativeList = null;

		currentLanguageWordItem = null;
		firstAssignmentWordItem = null;
		firstCollectionWordItem = null;
		firstContextWordItem = null;
		firstPossessiveNounWordItem = null;
		firstPredefinedWordItem = null;
		firstSpecificationWordItem = null;
		firstTouchedWordItem = null;
		firstUserProperNounWordItem = null;
		firstWordItem = null;
		lastCollectionWordItem = null;
		lastContextWordItem = null;
		lastPredefinedWordItem = null;
		predefinedNounLanguageWordItem = null;
		predefinedNounUserWordItem = null;

		// Don't initialize 'currentPathStringBuffer' there. 
		// It will be set in the main() method of the Thinknowlogy class.
		// Otherwise it will destroy the initialization by Thinknowlogy
//		currentPathStringBuffer = null;
		interfaceLanguageStringBuffer = null;
		learnedFromUserStringBuffer = null;
		queryStringBuffer = null;
		readUserSentenceStringBuffer = null;
		writtenSentenceStringBuffer = null;
		writtenUserSentenceStringBuffer = null;
		}
	};

/*************************************************************************
 *	"Honor the Lord for the glory of his name.
 *	Worship the Lord in the splendor of his holiness." (Psalm 29:2)
 *************************************************************************/

File: source/C++/GlobalVariables.cpp
/*	Class:			GlobalVariables
 *	Purpose:		To hold the global variables
 *	Version:		Thinknowlogy 2018r3 (Deep Magic)
 *************************************************************************/
/*	Copyright (C) 2009-2018, Menno Mafait. Your suggestions, modifications,
 *	corrections and bug reports are welcome at http://mafait.org/contact/
 *************************************************************************/
/*	This program is free software: you can redistribute it and/or modify
 *	it under the terms of the GNU General Public License as published by
 *	the Free Software Foundation, either version 2 of the License, or
 *	(at your option) any later version.
 *
 *	This program is distributed in the hope that it will be useful,
 *	but WITHOUT ANY WARRANTY; without even the implied warranty of
 *	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *	GNU General Public License for more details.
 *
 *	You should have received a copy of the GNU General Public License along
 *	with this program; if not, write to the Free Software Foundation, Inc.,
 *	51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *************************************************************************/

#include <string.h>
#include "Constants.h"

// Class declarations
class SelectionList;
class WordItem;

class GlobalVariables
	{
	friend class AdminImperative;
	friend class AdminItem;
	friend class AdminQuery;
	friend class AdminReadCreateWords;
	friend class AdminReadFile;
	friend class AdminReadSentence;
	friend class AdminReasoningNew;
	friend class AdminReasoningOld;
	friend class AdminSpecification;
	friend class AdminWrite;
	friend class CollectionItem;
	friend class CollectionList;
	friend class ContextItem;
	friend class ContextList;
	friend class FileItem;
	friend class FileList;
	friend class GeneralizationItem;
	friend class GeneralizationList;
	friend class GrammarItem;
	friend class GrammarList;
	friend class InputOutput;
	friend class InterfaceItem;
	friend class Item;
	friend class JustificationItem;
	friend class JustificationList;
	friend class List;
	friend class MultipleWordItem;
	friend class MultipleWordList;
	friend class ReadItem;
	friend class ReadList;
	friend class ScoreItem;
	friend class ScoreList;
	friend class SelectionItem;
	friend class SelectionList;
	friend class SpecificationItem;
	friend class SpecificationList;
	friend class WordItem;
	friend class WordList;
	friend class WordQuestion;
	friend class WordSpecification;
	friend class WordTypeItem;
	friend class WordTypeList;
	friend class WordWrite;
	friend class WriteItem;

	// Protected global variables

	bool hasDisplayedIntegrityWarning = false;
	bool hasDisplayedMessage = false;
	bool hasDisplayedWarning = false;
	bool hasFoundAnswerToQuestion = false;
	bool hasFoundQuery = false;
	bool isAssignmentChanged = false;
	bool isConflictingQuestion = false;
	bool isFirstAnswerToQuestion = false;
	bool isQuestionAlreadyAnswered = false;

	signed char result = RESULT_OK;

	unsigned short currentAssignmentLevel = NO_ASSIGNMENT_LEVEL;
	unsigned short currentLanguageNr = NO_LANGUAGE_NR;
	unsigned short currentUserNr = NO_USER_NR;
	unsigned short currentWriteLevel = NO_WRITE_LEVEL;

	unsigned int currentSentenceNr = 1;		// First sentence
	unsigned int currentSentenceItemNr = NO_ITEM_NR;

	unsigned int nActiveQueryItems = 0;
	unsigned int nInactiveQueryItems = 0;
	unsigned int nArchivedQueryItems = 0;
	unsigned int nReplacedQueryItems = 0;

	unsigned int nDeletedItems = 0;

	unsigned int nUserGeneralizationWords = 0;
	unsigned int nUserSpecificationWords = 0;
	unsigned int nUserRelationWords = 0;

	unsigned int removeSentenceNr = NO_SENTENCE_NR;
	unsigned int removeStartItemNr = NO_ITEM_NR;

	SelectionList *adminConditionList = NULL;
	SelectionList *adminActionList = NULL;
	SelectionList *adminAlternativeList = NULL;

	WordItem *currentLanguageWordItem = NULL;
	WordItem *firstAssignmentWordItem = NULL;
	WordItem *firstCollectionWordItem = NULL;
	WordItem *firstContextWordItem = NULL;
	WordItem *firstPossessiveNounWordItem = NULL;
	WordItem *firstPredefinedWordItem = NULL;
	WordItem *firstSpecificationWordItem = NULL;
	WordItem *firstTouchedWordItem = NULL;
	WordItem *firstUserProperNounWordItem = NULL;
	WordItem *firstWordItem = NULL;
	WordItem *lastCollectionWordItem = NULL;
	WordItem *lastContextWordItem = NULL;
	WordItem *lastPredefinedWordItem = NULL;
	WordItem *predefinedNounLanguageWordItem = NULL;
	WordItem *predefinedNounUserWordItem = NULL;

	char learnedFromUserString[MAX_SENTENCE_STRING_LENGTH] = EMPTY_STRING;
	char queryString[MAX_SENTENCE_STRING_LENGTH] = EMPTY_STRING;
	char writtenSentenceString[MAX_SENTENCE_STRING_LENGTH] = EMPTY_STRING;
	char writtenUserSentenceString[MAX_SENTENCE_STRING_LENGTH] = EMPTY_STRING;
	};

/*************************************************************************
 *	"Honor the Lord for the glory of his name.
 *	Worship the Lord in the splendor of his holiness." (Psalm 29:2)
 *************************************************************************/
